protocol/metrics/consumer_relayserver_client.go

```diff
@@ -0,0 +1,196 @@
+package metrics
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/lavanet/lava/utils"
+)
+
+type ConsumerRelayserverClient struct {
+	endPointAddress    string
+	addQueue           []UpdateMetricsRequest
+	sendQueue          []UpdateMetricsRequest
+	ticker             *time.Ticker
+	lock               sync.Mutex
+	sendID             int
+	isSendQueueRunning bool
+}
+type UpdateMetricsRequest struct {
+	RecordDate      string `json:"RecordDate"`
+	Hash            string `json:"Hash"`
+	Chain           string `json:"Chain"`
+	ApiType         string `json:"ApiType"`
+	RelaysInc       uint64 `json:"RelaysInc"`
+	CuInc           int    `json:"CuInc"`
+	LatencyToAdd    uint64 `json:"LatencyToAdd"`
+	LatencyAvgCount int    `json:"LatencyAvgCount"`
+}
+
+func NewConsumerRelayserverClient(endPointAddress string) *ConsumerRelayserverClient {
+	if endPointAddress == DisabledFlagOption {
+		utils.LavaFormatDebug("CUC: Consumer Usageserver disabled")
+		return nil
+	}
+
+	cuc := &ConsumerRelayserverClient{
+		endPointAddress: endPointAddress,
+		ticker:          time.NewTicker(30 * time.Second),
+		sendQueue:       make([]UpdateMetricsRequest, 0),
+		addQueue:        make([]UpdateMetricsRequest, 0),
+		lock:            sync.Mutex{},
+	}
+
+	utils.LavaFormatDebug("CUC: Starting processQueue goroutine")
+	go cuc.processQueue()
+
+	return cuc
+}
+
+func (cuc *ConsumerRelayserverClient) processQueue() {
+	if cuc == nil {
+		return
+	}
+
+	for range cuc.ticker.C {
+		cuc.lock.Lock()
+		if !cuc.isSendQueueRunning && len(cuc.addQueue) > 0 {
+			cuc.sendQueue, cuc.addQueue = cuc.addQueue, make([]UpdateMetricsRequest, 0)
+			cuc.isSendQueueRunning = true
+			cuc.sendID++
+			utils.LavaFormatDebug(fmt.Sprintf("CUC: Swapped queues (sendQueue length: %d) and set isSendQueueRunning to true - CCC iter:%d.", len(cuc.sendQueue), cuc.sendID))
+			cuc.lock.Unlock()
+
+			utils.LavaFormatDebug(fmt.Sprintf("CUC: Sending requests (sendQueue length: %d) - async start - CCC iter:%d.", len(cuc.sendQueue), cuc.sendID))
```

Comment:
reading from cuc while not locked is not allowed. 
better use of locks would be change the locks to a read write lock sync.RWMutex 
and read lock when can. reading is allowed in parallel by the lock itself so multiple reads are faster.

Reply:
done